v2/src/log_sink_console.c

```diff
@@ -1,18 +1,252 @@
+#include "c_logging/logger.h"
+
+#include "c_logging/log_sink_console.h" // IWYU pragma: keep
+
+#define MIN(a, b) ((a) < (b)) ? (a) : (b)
+
+/* Codes_SRS_LOG_SINK_CONSOLE_01_007: [ LOG_LEVEL_CRITICAL shall be displayed with bright red \x1b[31;1m. ]*/
+#define LOG_SINK_CONSOLE_ANSI_COLOR_BRIGHT_RED      "\x1b[31;1m"
+/* Codes_SRS_LOG_SINK_CONSOLE_01_008: [ LOG_LEVEL_ERROR shall be displayed with red \x1b[31m. ]*/
+#define LOG_SINK_CONSOLE_ANSI_COLOR_RED             "\x1b[31m"
+/* Codes_SRS_LOG_SINK_CONSOLE_01_009: [ LOG_LEVEL_WARNING shall be displayed with bright yellow \x1b[33;1m. ]*/
+#define LOG_SINK_CONSOLE_ANSI_COLOR_BRIGHT_YELLOW   "\x1b[33;1m"
+/* Codes_SRS_LOG_SINK_CONSOLE_01_010: [ LOG_LEVEL_INFO shall be displayed with yellow \x1b[33m. ]*/
+#define LOG_SINK_CONSOLE_ANSI_COLOR_YELLOW          "\x1b[33m"
+/* Codes_SRS_LOG_SINK_CONSOLE_01_011: [ LOG_LEVEL_VERBOSE shall be displayed with white \x1b[37m. ]*/
+#define LOG_SINK_CONSOLE_ANSI_COLOR_WHITE           "\x1b[37m"
+#define LOG_SINK_CONSOLE_ANSI_COLOR_RESET           "\x1b[0m"
+
+static const char* level_colors[] =
+{
+    LOG_SINK_CONSOLE_ANSI_COLOR_BRIGHT_RED,     // LOG_LEVEL_CRITICAL
+    LOG_SINK_CONSOLE_ANSI_COLOR_RED,            // LOG_LEVEL_ERROR
+    LOG_SINK_CONSOLE_ANSI_COLOR_BRIGHT_YELLOW,  // LOG_LEVEL_WARNING
+    LOG_SINK_CONSOLE_ANSI_COLOR_YELLOW,         // LOG_LEVEL_INFO
+    LOG_SINK_CONSOLE_ANSI_COLOR_WHITE,          // LOG_LEVEL_VERBOSE
+};
+
+const char error_string[] = "Error formatting log line\r\n";
+
+static int log_n_properties(char* buffer, size_t buffer_size, const LOG_CONTEXT_PROPERTY_VALUE_PAIR* property_value_pairs, size_t property_value_pair_count)
+{
+    int result = 0;
+
+    /* Codes_SRS_LOG_SINK_CONSOLE_01_016: [ For each property: ]*/
+    for (size_t i = 0; i < property_value_pair_count; i++)
+    {
+        /* Codes_SRS_LOG_SINK_CONSOLE_01_017: [ If the property type is struct (used as a container for context properties): ]*/
+        if (property_value_pairs[i].type->get_type() == LOG_CONTEXT_PROPERTY_TYPE_struct)
+        {
+            /* Codes_SRS_LOG_SINK_CONSOLE_01_025: [ log_sink_console.log_sink_log shall print the struct property name and an opening brace. ]*/
+            int snprintf_result = snprintf(buffer, buffer_size, " %s%s{", property_value_pairs[i].name, property_value_pairs[i].name[0] == 0 ? "" : "=");
+            if (snprintf_result < 0)
+            {
+                /* Codes_SRS_LOG_SINK_CONSOLE_01_022: [ If any encoding error occurs during formatting of the line (i.e. if any printf class functions fails), log_sink_console.log_sink_log shall print Error formatting log line and return. ]*/
+                result = -1;
+                break;
+            }
+            else
+            {
+                snprintf_result = MIN(snprintf_result, (int)buffer_size);
+                buffer += snprintf_result;
+                buffer_size -= snprintf_result;
+                result += snprintf_result;
```

Comment:
Wouldn't it be nice to move these three lines of code to a function that returns  snprintf_result/buffer_size and use it everywhere?

Reply:
Extracted in a macro.